sampler.go

```diff
@@ -184,9 +179,152 @@ func (s *rateLimitingSampler) Equal(other Sampler) bool {
 	return false
 }
 
-// Tags implements Tags of Sampler
-func (s *rateLimitingSampler) getTags() []tag {
-	return s.tags
+// -----------------------
+
+type lowerBoundSampler struct {
+	rateLimitingSampler Sampler
+	samplingRate        float64
+	tags                []Tag
+}
+
+// NewLowerBoundSampler creates a sampler that samples at most maxTracesPerSecond. It is a wrapper around
+// a rateLimitingSampler with different tags since it's intended to be used by the adaptiveSampler.
+// The sampler.param tag is sampling rate rather than maxTracesPerSecond. This is used to tell the collector
+// that the trace sampled by this sampler used the samplingRate to determine the throughput for the operation.
+func NewLowerBoundSampler(maxTracesPerSecond, samplingRate float64) (Sampler, error) {
+	tags := []Tag{
+		{key: SamplerTypeTagKey, value: SamplerTypeLowerBound},
+		{key: SamplerParamTagKey, value: samplingRate},
+	}
+	rateLimitingSampler, err := NewRateLimitingSampler(maxTracesPerSecond)
+	if err != nil {
+		return nil, err
+	}
+	return &lowerBoundSampler{
+		rateLimitingSampler: rateLimitingSampler,
+		samplingRate:        samplingRate,
+		tags:                tags,
+	}, nil
+}
+
+// IsSampled implements IsSampled() of Sampler.
+func (s *lowerBoundSampler) IsSampled(id uint64, operation string) (bool, []Tag) {
+	sampled, _ := s.rateLimitingSampler.IsSampled(id, operation)
+	return sampled, s.tags
+}
+
+func (s *lowerBoundSampler) Close() {
+	s.rateLimitingSampler.Close()
+}
+
+func (s *lowerBoundSampler) Equal(other Sampler) bool {
+	if o, ok := other.(*lowerBoundSampler); ok {
+		return s.samplingRate == o.samplingRate &&
+			s.rateLimitingSampler.Equal(o.rateLimitingSampler)
+	}
+	return false
+}
+
+// -----------------------
+
+type multiSampler struct {
+	probabilisticSampler Sampler
+	lowerBoundSampler    Sampler
+}
+
+type adaptiveSampler struct {
+	samplers map[string]*multiSampler
+
+	// TODO expand to keep track of newly seen operations
+	// TODO add a cap on the number of operations
+	// TODO need default rate limit and probabilistic rates for first time operations
+}
+
+// NewAdaptiveSampler adaptiveSampler is a delegating sampler that applies both probabilisticSampler and
+// rateLimitingSampler per operation. The probabilisticSampler is given higher priority when tags are emitted,
+// ie. if IsSampled() for both samplers return true, the tags for probabilisticSampler will be used.
+//
+// This sampler assumes that it will be used by RemotelyControlledSampler hence it is up to the user to
+// ensure this is thread-safe if not used by RemotelyControlledSampler.
+func NewAdaptiveSampler(maxTracesPerSecond float64, samplingRates map[string]*float64) (Sampler, error) {
```

Comment:
would it be useful to separate "adaptive" sampling from "per operation sampling"? The adaptive logic is non-trivial by itself, and easier to follow if it's not mixed up with the maintenance of per-operation mapping.